lib/collection/src/collection/resharding_update_pre_filter.rs

```diff
@@ -0,0 +1,96 @@
+use std::collections::HashSet;
+
+use segment::types::CustomIdCheckerCondition as _;
+
+use crate::hash_ring;
+use crate::operations::types::{CollectionResult, PointRequestInternal};
+use crate::operations::CollectionUpdateOperations;
+use crate::shards::resharding::ReshardStage;
+use crate::shards::shard::ShardId;
+use crate::shards::shard_holder::ShardHolder;
+use crate::shards::shard_trait::ShardOperation;
+
+#[derive(Clone, Debug)]
+pub enum ReshardingUpdatePreFilter {
+    NewShard,
+    OldShard(hash_ring::HashRingFilter),
+}
+
+impl ReshardingUpdatePreFilter {
+    pub fn new(shard_holder: &ShardHolder, shard_id: ShardId) -> Option<Self> {
+        let state = shard_holder.resharding_state()?;
+
+        if state.shard_id == shard_id && state.stage == ReshardStage::MigratingPoints {
+            Some(Self::NewShard)
+        } else if state.shard_id != shard_id && state.stage >= ReshardStage::ReadHashRingCommitted {
+            shard_holder.resharding_filter().map(Self::OldShard)
+        } else {
+            None
+        }
+    }
+
+    pub async fn apply(
+        &self,
+        operation: &mut CollectionUpdateOperations,
+        local_shard: &(dyn ShardOperation + Sync),
+        runtime: &tokio::runtime::Handle,
+    ) -> CollectionResult<()> {
+        // There's no need to pre-filter point delete operations
+        if operation.is_delete_points() {
+            return Ok(());
+        }
+
+        let points_to_filter = match self {
```

Comment:
Points that are not moved as part of resharding should not be affected at all.

In other words, we should only check the condition on point IDs here that are distributed to another shard according to the hash rings. All other point IDs should be kept no matter what. I believe we had a `is_point_moved` (or something) helper for that.

Reply:
> Points that are not moved as part of resharding should not be affected at all.

That's exactly how it works already? Can you clarify, cause I can't see what's wrong currently.

Reply:
I think I got it. It matters when pre-filtering on receiver shard during resharding down. 🤦‍♀️